datafusion/physical-plan/src/joins/nested_loop_join.rs

```diff
@@ -560,91 +729,148 @@ impl NestedLoopJoinStream {
+                        }
+                    }
+                    Poll::Pending => {
+                        return match self.output_buffer.flush() {
+                            Ok(Some(batch)) => {
+                                // If there was anything in the output buffer flush it
+                                // so that it can be processed.
+                                self.join_metrics.output_batches.add(1);
+                                self.join_metrics.output_rows.add(batch.num_rows());
+                                Poll::Ready(Some(Ok(batch)))
+                            }
+                            Ok(None) => Poll::Pending,
+                            Err(err) => Poll::Ready(Some(Err(err))),
                         };
+                    }
+                }
+            }
 
-                        // Only setting up timer, input is exhausted
-                        let timer = self.join_metrics.join_time.timer();
-                        // use the global left bitmap to produce the left indices and right indices
-                        let (left_side, right_side) =
-                            get_final_indices_from_shared_bitmap(
-                                visited_left_side,
-                                self.join_type,
-                            );
-                        let empty_right_batch =
-                            RecordBatch::new_empty(self.outer_table.schema());
-                        // use the left and right indices to produce the batch result
-                        let result = build_batch_from_indices(
-                            &self.schema,
-                            left_data.batch(),
-                            &empty_right_batch,
-                            &left_side,
-                            &right_side,
-                            &self.column_indices,
-                            JoinSide::Left,
-                        );
-                        self.is_exhausted = true;
-
-                        // Recording time & updating output metrics
-                        if let Ok(batch) = &result {
-                            timer.done();
-                            self.join_metrics.output_batches.add(1);
-                            self.join_metrics.output_rows.add(batch.num_rows());
-                        }
+            debug_assert!(self.outer_record_batch.is_some());
+            let right_batch = self.outer_record_batch.as_ref().unwrap();
+            let num_rows = if left_data.batch().num_rows() == 0 {
+                self.output_buffer.needed_rows()
+            } else {
+                std::cmp::max(
+                    1,
+                    self.output_buffer.needed_rows() / left_data.batch().num_rows(),
+                )
+            };
+            let num_rows = std::cmp::min(
+                num_rows,
+                right_batch.num_rows() - self.outer_record_batch_row,
+            );
+            let sliced_right_batch =
```

Comment:
This seems to be the core concern / potential loss of performance (joining one row at a time)

An alternate solution might be to
1. Try to join the entire batch, but return an error if the number of matched rows exceeds some threshold (the target batch size? 2x the target batchsize?)
2. If the "too many matched rows" error is returned, then fallback to joining one row at a time

This is indeed more complicated, but it would mean we only pay the performance hit when a large number of matching rows are produced (aka when the join is increasing cardinality significantly)